### ISSUE TITLE
Add IntX and FloatX Converting Clamp Methods

### ISSUE THREAD

**Issue by RespiteSage** [crystal-lang/crystal#10561]
## Feature Request

- Describe considered alternative solutions, and the reasons why you have not proposed them as a solution here.
- Does it break backward compatibility, if yes then what's the migration path?

## Motivation

`Comparable#clamp` is fantastic, and it works very well. However, its signature allows any types for the clamp boundaries (implicitly requiring that `self` defines `#<` against the boundary types), and it returns the boundary directly if the boundary is non-nil and `self` is "outside" the clamping boundary.

This explanation is basically exactly the current implementation:

```crystal
def clamp(min, max)
  return max if !max.nil? && self > max
  return min if !min.nil? && self < min
  self
end
```
[source](https://github.com/crystal-lang/crystal/blob/dd40a2442/src/comparable.cr#L21)

I was recently messing with `Float64#clamp`, and found that it returns a union when the boundaries are integer literals:

```crystal
a = (5.0).clamp(0,10)

puts typeof(a) # => (Float64 | Int32)
```

What I expected and desired was that the return type of `#clamp` would be the type of the value that was clamped.

## Proposed Feature

I propose that we add custom clamping methods to each `IntX` and `FloatX` type. However, the `Int` subtype methods would have slightly different behavior from the `Float` subtype methods.

The `Int` subtype methods would take other integer types as the boundaries and, if `self` is outside of a boundary, return the boundary converted to `self`'s type. So, for instance, for `Int32`:

```crystal
def clamp(min : Int?, max : Int?) : Int32
  return max.to_i32 if !max.nil? && self > max
  return min.to_i32 if !min.nil? && self < min
  self
end
```
(Note: I'm retaining the nil semantics from the normal implementation, but I'm not mentioning them in my overall argument, since they should be irrelevant.)

I don't think we have to worry about overflow here if `#clamp` is called correctly because the boundary check will restrict the `#to_X` call to values that will fit in that type. I suppose that overflow is possible if you pass in a `min` that is larger than `max`, though.

`Float` subtypes, on the other hand, can be assigned to integer values as well, so they will take `Number` boundaries and return the boundary converted to `self`'s type. So, for instance, for `Float64`:

```crystal
def clamp(min : Number?, max : Number?) : Float64
  return max.to_f32 if !max.nil? && self > max
  return min.to_f32 if !min.nil? && self < min
  self
end
```

## Problems with the Proposed Feature

The most obvious problem with the proposal is that it removes the possibility of intentionally using `#clamp` to change the type of a numeric value to a different numeric type. I don't know if this is an intended use case, but removing that behavior *is* a breaking change (a much bigger concern now that we're past 1.0).

The other problem is that there's a question of whether this is worthwhile at all. In my example at the beginning of this issue, I could just change the boundaries to floating point literals, and everything will work exactly as desired. That's not a huge hassle, and while the proposed implementation is very straightforward, it's also quite a lot of code overall that would not only need to be written but also need to be maintained.

## Conclusion

I do think this would be a good standard library "ergonomics"/convenience feature, but I want to know what others think. The downsides aren't huge, but I'm not sure how widely useful this would be, either.

**Comment by straight-shoota** [comment]
Yes, I think it should be expected that `T#clamp(X, Y)` returns `T`. `X` or `Y` are just boundaries meant for comparison. We can't enforce this generically for any type that is comparable with other types. But at least for number types this is easy to do.

The implementation can make use of `self.class.new(min)` for a generic solution. So no need for individual methods per type.

**Comment by asterite** [comment]
What if you clamp an int between two floats?

**Comment by straight-shoota** [comment]
The motivation is to remove unintended union types. I suppose an alternative approach would be to cast the range values to the clamped type at the call side.
So, yeah maybe this doesn't need action.

Essentially it comes down to detailed semantics of a call like `1.clamp(1.5, 3)`:
Does it mean restricting a value on the scale of the data type to fit in the given range? Or does it mean to restrict a value to the given range and map out-of range values to the range boundaries? The current behaviour is the latter, but I'd assume that's mostly an implementation detail.

Is there any use case for the behaviour that in-range values and out-of-range values are different data types?

**Comment by asterite** [comment]
This seems to work fine:

```crystal
module Comparable(T)
  def clamp(min : Number, max : Number)
    result_type = typeof(self + min + max)

    return result_type.new(max) if !max.nil? && self > max
    return result_type.new(min) if !min.nil? && self < min
    result_type.new(self)
  end
end

p!(
  typeof(1.clamp(2, 3)),
  typeof(1.5.clamp(2, 3)),
  typeof(1.5.clamp(2, 3.5)),
  typeof(1.clamp(2.5, 3)),
  typeof(1.clamp(2.5, 3.5)),
  typeof("good".clamp("hello", "wrong"))
)
```

Output:

```
typeof(1.clamp(2, 3))                  # => Int32
typeof(1.5.clamp(2, 3))                # => Float64
typeof(1.5.clamp(2, 3.5))              # => Float64
typeof(1.clamp(2.5, 3))                # => Float64
typeof(1.clamp(2.5, 3.5))              # => Float64
typeof("good".clamp("hello", "wrong")) # => String
```

**Comment by asterite** [comment]
Another alternative is to define clamp like `def clamp(min : self, max : self)` so there's no confusion: you always have to pass the same type everywhere.

**Comment by straight-shoota** [comment]
Uses the already established type promotion strategy for the addition operator with `typeof(self + min + max)` is pretty neat. That looks like the best solution.

**Comment by asterite** [comment]
It might be good to define this just in Number, just in case.
